community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/combined/CombinedSchemaIndexProvider.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2002-2017 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.index.schema.combined;
+
+import java.io.IOException;
+
+import org.neo4j.io.fs.FileSystemAbstraction;
+import org.neo4j.io.pagecache.PageCache;
+import org.neo4j.kernel.api.index.IndexAccessor;
+import org.neo4j.kernel.api.index.IndexPopulator;
+import org.neo4j.kernel.api.index.InternalIndexState;
+import org.neo4j.kernel.api.index.SchemaIndexProvider;
+import org.neo4j.kernel.api.schema.index.IndexDescriptor;
+import org.neo4j.kernel.impl.api.index.sampling.IndexSamplingConfig;
+import org.neo4j.kernel.impl.storemigration.StoreMigrationParticipant;
+import org.neo4j.storageengine.api.schema.IndexSample;
+import org.neo4j.values.storable.Value;
+import org.neo4j.values.storable.ValueGroup;
+
+import static java.lang.String.format;
+
+/**
+ * Exactly what this provider will be called and where it will live is undecided, so for the time being
+ * it will be called something temporary.
+ *
+ * The idea is to have a boosted provider which can handle some type of data and is faster, where all other
+ * data is managed by the fallback provider.
+ */
+public class CombinedSchemaIndexProvider extends SchemaIndexProvider
+{
+    private final SchemaIndexProvider boostProvider;
+    private final SchemaIndexProvider fallbackProvider;
+
+    public CombinedSchemaIndexProvider( SchemaIndexProvider boostProvider, SchemaIndexProvider fallbackProvider )
```

Comment:
There are a few alternatives:

1. Lucene and Combined providers living side by side.
What provider is used when an index is created will then look something like:
```
if ( compositeIndex ) { return LuceneProvider }
if ( configuredToOnlyUseLucene ) { return LuceneProvider }
return CombinedProvider
```
In this way, the Combined index will only be used when there is actually a chance that it will see updates for Native part.

2. We only have Combined provider (like what this PR assumes)
Then it will need to handle the case for composite keys as well (by sending them all to Lucene).

What route we will take is still undecided and the this code will adapt when a decision is taken.

Did this make it more clear?